### DISCUSSION TITLE
Onnx file Not Found

### DISCUSSION THREAD

**Discussion by mushie4282** [dscripka/openWakeWord#82] [Q&A]
Not sure if anyone had the same issue, but in the last cell of `automatic_model_training_simple.ipynb`, I get this error `FileNotFoundError: [Errno 2] No such file or directory: 'my_custom_model/Hey_ARCKK.onnx` after training is finished.

Context:
Running on Windows 11 machine
I have restarted the process multiple times. 

I appreciate any help or tips. Thank you in advance.

**Comment by jesgarcas** [comment]
I'm in the same situation

```
FileNotFoundError                         Traceback (most recent call last)

[<ipython-input-8-ab8a2170a404>](https://localhost:8080/#) in <cell line: 92>()
     90     return None
     91 
---> 92 convert_onnx_to_tflite(f"my_custom_model/{config['model_name']}.onnx", f"my_custom_model/{config['model_name']}.tflite")
     93 
     94 # Automatically download the trained model files

2 frames

[/usr/local/lib/python3.10/dist-packages/onnx/__init__.py](https://localhost:8080/#) in _load_bytes(f)
    143     else:
    144         f = typing.cast(Union[str, os.PathLike], f)
--> 145         with open(f, "rb") as readable:
    146             content = readable.read()
    147     return content

FileNotFoundError: [Errno 2] No such file or directory: 'my_custom_model/example.onnx'
```

**Comment by dscripka** [comment]
@mushie4282 and @jesgarcas, I've likely found the cause of these issues, have pushed fixes to openWakeWord, and updated the training Colab notebooks accordingly. Please try the latest versions of the Colab notebooks, and let me know if you have any more problems.

**Reply by mushie4282** [reply]
Thank you for updating the Colab notebooks. I just tried the `automatic_model_training_simple.ipynb` and I'm getting the same error. 
```warnings.warn(
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
[<ipython-input-3-ab8a2170a404>](https://localhost:8080/#) in <cell line: 92>()
     90     return None
     91 
---> 92 convert_onnx_to_tflite(f"my_custom_model/{config['model_name']}.onnx", f"my_custom_model/{config['model_name']}.tflite")
     93 
     94 # Automatically download the trained model files

2 frames
[/usr/local/lib/python3.10/dist-packages/onnx/__init__.py](https://localhost:8080/#) in _load_bytes(f)
    143     else:
    144         f = typing.cast(Union[str, os.PathLike], f)
--> 145         with open(f, "rb") as readable:
    146             content = readable.read()
    147     return content
FileNotFoundError: [Errno 2] No such file or directory: 'my_custom_model/hey_ARCKK.onnx'

**Reply by dscripka** [reply]
@mushie4282, can you share any other errors that might have happened in other parts of the notebook? From my testing this error should have been resolved with the latest updates.

**Reply by mushie4282** [reply]
Hi, I was able to successfully run the `automatic_model_training_simple.ipynb` notebook on a new Google account. I'm not sure what the issue was on my previous account, but I'm glad it's working now. Thank you for your help!

**Comment by CagoBHuK** [comment]
I've got the same issue under two diffirent Google users.
Такая же проблема под двумя разными пользователями Гугл. Оперативно отвечу на вопросы в телеге @CagoBHuK.


`Traceback (most recent call last):
  File "/content/openwakeword/openwakeword/train.py", line 888, in <module>
    best_model = oww.auto_train(
  File "/content/openwakeword/openwakeword/train.py", line 276, in auto_train
    self.train_model(
  File "/content/openwakeword/openwakeword/train.py", line 447, in train_model
    for step_ndx, data in tqdm(enumerate(X, 0), total=max_steps, desc="Training"):
  File "/usr/local/lib/python3.10/dist-packages/tqdm/std.py", line 1182, in __iter__
    for obj in iterable:
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 630, in __next__
    data = self._next_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1345, in _next_data
    return self._process_data(data)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1371, in _process_data
    data.reraise()
  File "/usr/local/lib/python3.10/dist-packages/torch/_utils.py", line 694, in reraise
    raise exception
ValueError: Caught ValueError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 41, in fetch
    data = next(self.dataset_iter)
  File "/content/openwakeword/openwakeword/data.py", line 852, in __next__
    return np.vstack(X), np.array(y)
  File "<__array_function__ internals>", line 180, in vstack
  File "/usr/local/lib/python3.10/dist-packages/numpy/core/shape_base.py", line 282, in vstack
    return _nx.concatenate(arrs, 0)
  File "<__array_function__ internals>", line 180, in concatenate
ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 16 and the array at index 1 has size 20

/usr/local/lib/python3.10/dist-packages/tensorflow_addons/utils/tfa_eol_msg.py:23: UserWarning: 

TensorFlow Addons (TFA) has ended development and introduction of new features.
TFA has entered a minimal maintenance and release mode until a planned end of life in May 2024.
Please modify downstream libraries to take dependencies from other repositories in our TensorFlow community (e.g. Keras, Keras-CV, and Keras-NLP). 

For more information see: https://github.com/tensorflow/addons/issues/2807 

  warnings.warn(
/usr/local/lib/python3.10/dist-packages/tensorflow_addons/utils/ensure_tf_install.py:53: UserWarning: Tensorflow Addons supports using Python ops for all Tensorflow versions above or equal to 2.13.0 and strictly below 2.16.0 (nightly versions are not supported). 
 The versions of TensorFlow you are currently using is 2.8.1 and is not supported. 
Some things might work, some things might not.
If you were to encounter a bug, do not file an issue.
If you want to make sure you're using a tested and supported configuration, either change the TensorFlow version or the TensorFlow Addons's version. 
You can find the compatibility matrix in TensorFlow Addon's readme:
https://github.com/tensorflow/addons
  warnings.warn(
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
[<ipython-input-8-ab8a2170a404>](https://localhost:8080/#) in <cell line: 92>()
     90     return None
     91 
---> 92 convert_onnx_to_tflite(f"my_custom_model/{config['model_name']}.onnx", f"my_custom_model/{config['model_name']}.tflite")
     93 
     94 # Automatically download the trained model files

2 frames
[/usr/local/lib/python3.10/dist-packages/onnx/__init__.py](https://localhost:8080/#) in _load_bytes(f)
    143     else:
    144         f = typing.cast(Union[str, os.PathLike], f)
--> 145         with open(f, "rb") as readable:
    146             content = readable.read()
    147     return content

FileNotFoundError: [Errno 2] No such file or directory: 'my_custom_model/krasnaia_karalaeva.onnx'`

**Comment by CagoBHuK** [comment]
On the second step I've got the error: ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pandas-gbq 0.19.2 requires google-auth-oauthlib>=0.7.0, but you have google-auth-oauthlib 0.4.6 which is incompatible.

**Comment by CagoBHuK** [comment]
Another one error at stage 2. As I see, pip can't install onnx_tf.

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
dopamine-rl 4.0.6 requires tensorflow>=2.2.0, which is not installed.

**Comment by CagoBHuK** [comment]
Edited deps install to
```

# install other dependencies
!pip install google-auth-oauthlib==0.4.6
!pip install mutagen==1.47.0
!pip install torchinfo==1.8.0
!pip install torchmetrics==1.2.0
!pip install speechbrain==0.5.14
!pip install audiomentations==0.33.0
!pip install torch-audiomentations==0.11.0
!pip install acoustics==0.2.6
!pip install tensorflow==2.8.1
!pip install tensorflow-cpu==2.8.1
!pip install tensorflow_probability==0.16.0
!pip install onnx_tf==1.10.0
!pip install pronouncing==0.2.0
!pip install datasets==2.14.6
!pip install deep-phonemizer==0.0.19
```

No deps errors. But still get: "ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 16 and the array at index 1 has size 20," - error at training stage.

**Comment by dscripka** [comment]
@CagoBHuK I just tried the [3 step training notebook](https://colab.research.google.com/drive/1q1oe2zOyZp7UsB3jJiQ1IFn8z5YfjwEb) linked in the readme, and the all 3 steps completed correctly without errors, producing both the `onnx` and `tflite` model files. I did see the warnings in step 2 about pip compatibility errors, but they seemed to be just warnings that did not impact the rest of the process.

However, the fact that you are getting this error:

```"ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 16 and the array at index 1 has size 20,"```

indicates that something about your wake word may be related. Can you share the exact wakeword you are using?

**Reply by CagoBHuK** [reply]
krasnaia_karalaeva

**Reply by dscripka** [reply]
@CagoBHuK this is a very late response (apologies!), but I just pushed a [commit](https://github.com/dscripka/openWakeWord/commit/c40fe924ffa12e9ddf24a3e5fcdeb4fd58ab07eb) that should fix this issue. If you pull the latest updates from `main` you can try again and see if you can successfully train with the "krasnaia_karalaeva" wakeword.

**Comment by CagoBHuK** [comment]
> @CagoBHuK this is a very late response (apologies!), but I just pushed a [commit](https://github.com/dscripka/openWakeWord/commit/c40fe924ffa12e9ddf24a3e5fcdeb4fd58ab07eb) that should fix this issue. If you pull the latest updates from `main` you can try again and see if you can successfully train with the "krasnaia_karalaeva" wakeword.

You did it! It now works for me! Thank you!

**Comment by Mandarzzz** [comment]
{
	"name": "ValueError",
	"message": "Arg specs do not match: original=FullArgSpec(args=['input', 'dtype', 'name', 'layout'], varargs=None, varkw=None, defaults=(None, None, None), kwonlyargs=[], kwonlydefaults=None, annotations={}), new=FullArgSpec(args=['input', 'dtype', 'name'], varargs=None, varkw=None, defaults=(None, None), kwonlyargs=[], kwonlydefaults=None, annotations={}), fn=<function ones_like_v2 at 0x7f625c5e28b0>",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[33], line 28
     24             f.write(tflite_model)
     26     return None
---> 28 convert_onnx_to_tflite(f\"my_custom_model/{config['model_name']}.onnx\", f\"my_custom_model/{config['model_name']}.tflite\")

Cell In[33], line 11, in convert_onnx_to_tflite(onnx_model_path, output_path)
      9 import logging
     10 import tempfile
---> 11 from onnx_tf.backend import prepare
     12 import tensorflow as tf
     14 # Convert to tflite from onnx model

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/onnx_tf/__init__.py:1
----> 1 from . import backend
      2 from .version import version as __version__

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/onnx_tf/backend.py:28
     26 from onnx_tf.common import get_unique_suffix
     27 from onnx_tf.common import supports_device as common_supports_device
---> 28 from onnx_tf.common.handler_helper import get_all_backend_handlers
     29 from onnx_tf.pb_wrapper import OnnxNode
     30 from onnx_tf.backend_tf_module import BackendTFModule, TFModule

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/onnx_tf/common/handler_helper.py:3
      1 from onnx import defs
----> 3 from onnx_tf.handlers.backend import *  # noqa
      4 from onnx_tf.handlers.backend_handler import BackendHandler
      5 import onnx_tf.common as common

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/onnx_tf/handlers/backend/bernoulli.py:2
      1 import tensorflow as tf
----> 2 from tensorflow_probability import distributions as tfd
      4 from onnx_tf.handlers.backend_handler import BackendHandler
      5 from onnx_tf.handlers.handler import onnx_op

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/__init__.py:23
     20 from tensorflow_probability import substrates
     21 # from tensorflow_probability.google import staging  # DisableOnExport
     22 # from tensorflow_probability.google import tfp_google  # DisableOnExport
---> 23 from tensorflow_probability.python import *  # pylint: disable=wildcard-import
     24 from tensorflow_probability.python.version import __version__

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/__init__.py:138
    135 if _tf_loaded():
    136   # Non-lazy load of packages that register with tensorflow or keras.
    137   for pkg_name in _maybe_nonlazy_load:
--> 138     dir(globals()[pkg_name])  # Forces loading the package from its lazy loader.
    141 all_util.remove_undocumented(__name__, _lazy_load + _maybe_nonlazy_load)

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/internal/lazy_loader.py:57, in LazyLoader.__dir__(self)
     56 def __dir__(self):
---> 57   module = self._load()
     58   return dir(module)

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/internal/lazy_loader.py:40, in LazyLoader._load(self)
     38   self._on_first_access = None
     39 # Import the target module and insert it into the parent's namespace
---> 40 module = importlib.import_module(self.__name__)
     41 if self._parent_module_globals is not None:
     42   self._parent_module_globals[self._local_name] = module

File ~/miniconda3/envs/myenv/lib/python3.9/importlib/__init__.py:127, in import_module(name, package)
    125             break
    126         level += 1
--> 127 return _bootstrap._gcd_import(name[level:], package, level)

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/experimental/__init__.py:31
     15 \"\"\"TensorFlow Probability API-unstable package.
     16 
     17 This package contains potentially useful code which is under active development
   (...)
     27 You are welcome to try any of this out (and tell us how well it works for you!).
     28 \"\"\"
     30 from tensorflow_probability.python.experimental import auto_batching
---> 31 from tensorflow_probability.python.experimental import bijectors
     32 from tensorflow_probability.python.experimental import distribute
     33 from tensorflow_probability.python.experimental import distributions

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/experimental/bijectors/__init__.py:17
      1 # Copyright 2020 The TensorFlow Probability Authors.
      2 #
      3 # Licensed under the Apache License, Version 2.0 (the \"License\");
   (...)
     13 # limitations under the License.
     14 # ============================================================================
     15 \"\"\"TensorFlow Probability experimental bijectors package.\"\"\"
---> 17 from tensorflow_probability.python.bijectors.ldj_ratio import forward_log_det_jacobian_ratio
     18 from tensorflow_probability.python.bijectors.ldj_ratio import inverse_log_det_jacobian_ratio
     19 from tensorflow_probability.python.experimental.bijectors.distribution_bijectors import make_distribution_bijector

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/bijectors/__init__.py:19
     15 \"\"\"Bijective transformations.\"\"\"
     17 # pylint: disable=unused-import,wildcard-import,line-too-long,g-importing-member
---> 19 from tensorflow_probability.python.bijectors.absolute_value import AbsoluteValue
     20 from tensorflow_probability.python.bijectors.ascending import Ascending
     21 from tensorflow_probability.python.bijectors.batch_normalization import BatchNormalization

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/bijectors/absolute_value.py:19
     15 \"\"\"AbsoluteValue bijector.\"\"\"
     17 import tensorflow.compat.v2 as tf
---> 19 from tensorflow_probability.python.bijectors import bijector
     20 from tensorflow_probability.python.internal import assert_util
     21 from tensorflow_probability.python.internal import dtype_util

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/bijectors/bijector.py:26
     24 from tensorflow_probability.python.internal import assert_util
     25 from tensorflow_probability.python.internal import auto_composite_tensor
---> 26 from tensorflow_probability.python.internal import batch_shape_lib
     27 from tensorflow_probability.python.internal import cache_util
     28 from tensorflow_probability.python.internal import dtype_util

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/internal/batch_shape_lib.py:23
     20 import numpy as np
     21 import tensorflow.compat.v2 as tf
---> 23 from tensorflow_probability.python.internal import prefer_static as ps
     24 from tensorflow_probability.python.internal import tensor_util
     25 from tensorflow_probability.python.internal import tensorshape_util

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/internal/prefer_static.py:344
    342     return np.ones(s_, dtype_util.as_numpy_dtype(dtype or input.dtype))
    343   return tf.ones(s, dtype or input.dtype, name)
--> 344 ones_like = _copy_docstring(tf.ones_like, _ones_like)
    347 def _rank(input, name=None):  # pylint: disable=redefined-builtin,unused-argument
    348   if not hasattr(input, 'shape'):

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/tensorflow_probability/python/internal/prefer_static.py:91, in _copy_docstring(original_fn, new_fn)
     89 new_spec = tf_inspect.getfullargspec(new_fn)
     90 if original_spec != new_spec:
---> 91   raise ValueError(
     92       'Arg specs do not match: original={}, new={}, fn={}'.format(
     93           original_spec, new_spec, original_fn))
     94 @decorator.decorator
     95 def wrap(wrapped_fn, *args, **kwargs):
     96   del wrapped_fn

ValueError: Arg specs do not match: original=FullArgSpec(args=['input', 'dtype', 'name', 'layout'], varargs=None, varkw=None, defaults=(None, None, None), kwonlyargs=[], kwonlydefaults=None, annotations={}), new=FullArgSpec(args=['input', 'dtype', 'name'], varargs=None, varkw=None, defaults=(None, None), kwonlyargs=[], kwonlydefaults=None, annotations={}), fn=<function ones_like_v2 at 0x7f625c5e28b0>"
}

I am getting above Error I have install ubuntu  working on it 
!pip install tensorflow==2.8.1
!pip install tensorflow-cpu==2.8.1
!pip install tensorflow_probability==0.16.0
!pip install onnx_tf==1.10.0

all the 4 dependency I check its all okay not able to convert the model in tf or onnx
